Zilon.Core/Zilon.Core/World/GlobeTransitionHandler.cs

```diff
@@ -0,0 +1,98 @@
+﻿using System;
+using System.Linq;
+using System.Threading;
+using System.Threading.Tasks;
+
+using Zilon.Core.MapGenerators;
+using Zilon.Core.Tactics;
+
+namespace Zilon.Core.World
+{
+    public sealed class GlobeTransitionHandler : IGlobeTransitionHandler, IDisposable
+    {
+        private readonly IGlobeExpander _globeExpander;
+
+        private readonly SemaphoreSlim _semaphoreSlim;
+
+        public GlobeTransitionHandler(IGlobeExpander globeExpander)
+        {
+            _globeExpander = globeExpander ?? throw new ArgumentNullException(nameof(globeExpander));
+
+            //Instantiate a Singleton of the Semaphore with a value of 1. This means that only 1 thread can be granted access at a time.
+            _semaphoreSlim = new SemaphoreSlim(1, 1);
+        }
+
+        public Task ProcessAsync(IGlobe globe, ISector sector, IActor actor, RoomTransition transition)
+        {
+            if (globe is null)
+            {
+                throw new ArgumentNullException(nameof(globe));
+            }
+
+            if (sector is null)
+            {
+                throw new ArgumentNullException(nameof(sector));
+            }
+
+            if (actor is null)
+            {
+                throw new ArgumentNullException(nameof(actor));
+            }
+
+            if (transition is null)
+            {
+                throw new ArgumentNullException(nameof(transition));
+            }
+
+            return ProcessInnerAsync(globe, sector, actor, transition);
+        }
+
+        private async Task ProcessInnerAsync(IGlobe globe, ISector sector, IActor actor, RoomTransition transition)
+        {
+            var sectorNode = transition.SectorNode;
+
+            //TODO Разобраться с этим кодом.
+            // https://blog.cdemi.io/async-waiting-inside-c-sharp-locks/
+            //Asynchronously wait to enter the Semaphore. If no-one has been granted access to the Semaphore, code execution will proceed, otherwise this thread waits here until the semaphore is released 
+#pragma warning disable CA2007 // Consider calling ConfigureAwait on the awaited task
+            await _semaphoreSlim.WaitAsync();
+#pragma warning restore CA2007 // Consider calling ConfigureAwait on the awaited task
+            try
+            {
+                if (sectorNode.State != SectorNodeState.SectorMaterialized)
+                {
+                    await _globeExpander.ExpandAsync(sectorNode).ConfigureAwait(false);
+                    globe.AddSectorNode(sectorNode);
+                }
+
+                try
+                {
+                    sector.ActorManager.Remove(actor);
+                }
+                catch (InvalidOperationException exception)
+                {
+                    // Пока ничего не делаем
+                    Console.WriteLine(exception);
```

Comment:
Codacy found an issue: [Remove this logging statement.](https://app.codacy.com/manual/kreghek/Zilon_Roguelike/pullRequest?prid=6105716)